oximeter/instruments/src/kstat/sampler.rs

```diff
@@ -513,149 +443,261 @@ impl KstatSamplerWorker {
+                    "id" => ?id,
+                );
+                let response = match self.targets.get(&id) {
+                    None => Err(Error::NoSuchTarget),
+                    Some(SampledObject::Kstat(k)) => Ok(TargetStatus::Ok {
+                        last_collection: k.time_of_last_collection,
+                    }),
+                    Some(SampledObject::Expired(e)) => {
+                        Ok(TargetStatus::Expired {
+                            reason: e.reason,
+                            error: e.error.to_string(),
+                            expired_at: e.expired_at,
+                        })
+                    }
+                };
+                match reply_tx.send(response) {
+                    Ok(_) => trace!(self.log, "sent reply"),
+                    Err(e) => error!(
+                        self.log,
+                        "failed to send reply";
+                        "id" => ?id,
+                        "error" => ?e,
+                    ),
+                }
+                None
+            }
+            #[cfg(all(test, target_os = "illumos"))]
+            Request::CreationTimes { reply_tx } => {
+                debug!(self.log, "request for creation times");
+                reply_tx.send(self.creation_times.clone()).unwrap();
+                debug!(self.log, "sent reply for creation times");
+                None
+            }
+        }
+    }
+
+    fn sample_next_target(
+        &mut self,
+        id: TargetId,
+        interval: Duration,
+        #[cfg(all(test, target_os = "illumos"))]
+        sample_count_tx: &mpsc::UnboundedSender<SampleCounts>,
+    ) -> Option<YieldIdAfter> {
+        match self.sample_one(id) {
+            Ok(Some(samples)) => {
+                if samples.is_empty() {
+                    debug!(
+                        self.log,
+                        "no new samples from target, requeueing";
+                        "id" => ?id,
+                    );
+                    return Some(YieldIdAfter::new(id, interval));
+                }
+                let n_samples = samples.len();
+                debug!(
+                    self.log,
+                    "pulled samples from target";
+                    "id" => ?id,
+                    "n_samples" => n_samples,
+                );
+
+                // Append any samples to the per-target queues.
+                //
+                // This returns None if there was no queue, and logs
+                // the error internally. We'll just go round the
+                // loop again in that case.
+                let Some(n_overflow_samples) =
+                    self.append_per_target_samples(id, samples)
+                else {
+                    return None;
+                };
+
+                // Safety: We only get here if the `sample_one()`
+                // method works, which means we have a record of the
+                // target, and it's not expired.
```

Comment:
silly nit, considered pedantic: i realize this code was already here, and it just moved around in this diff, but _personally_ i don't really love the use of a "Safety:" comment on something that isn't an `unsafe` block. if it were me, i might just put this text in the panic message for the `unreachable!` macro instead. feel free to disregard this, it's definitely not important.